tachyon/crypto/commitments/fri/two_adic_fri_pcs.h

```diff
@@ -0,0 +1,403 @@
+    math::RowMajorMatrix<F>& lde = mmcs_.GetMatrices(prover_data)[idx];
+    CHECK_GE(lde.rows(), coset.domain()->size());
+    Eigen::Block<math::RowMajorMatrix<F>> ret =
+        lde.topRows(coset.domain()->size());
+    ReverseMatrixIndexBits(ret);
+    return ret;
+  }
+
+  [[nodiscard]] bool CreateOpeningProof(
+      const std::vector<ProverData>& prover_data_by_round,
+      const std::vector<Points>& points_by_round, Challenger challenger,
+      OpenedValues* openings, TwoAdicFriProof* proof) {
+    ExtF alpha = challenger.template SampleExtElement<ExtF>();
+    size_t num_rounds = prover_data_by_round.size();
+
+    std::vector<std::vector<math::RowMajorMatrix<F>>> matrices_by_round(
+        num_rounds);
+    size_t global_max_num_rows = 0;
+    for (size_t round = 0; round < num_rounds; ++round) {
+      const ProverData& prover_data = prover_data_by_round[round];
+      matrices_by_round[round] = mmcs_.GetMatrices(prover_data);
+      global_max_num_rows =
+          std::max(global_max_num_rows, mmcs_.GetMaxRowSize(prover_data));
+    }
+    size_t log_global_max_num_rows =
+        base::bits::CheckedLog2(global_max_num_rows);
+
+    // For each unique opening point z, we will find the largest degree bound
+    // for that point, and precompute 1/(X - z) for the largest subgroup (in
+    // bitrev order).
+    absl::flat_hash_map<ExtF, std::vector<ExtF>> inv_denoms =
+        ComputeInverseDenominators(
+            matrices_by_round, points_by_round,
+            F::FromMontgomery(F::Config::kSubgroupGenerator));
+
+    std::array<std::vector<ExtF>, 32> reduced_openings;
+    std::array<size_t, 32> num_reduced = {};
+
+    OpenedValues opened_values(num_rounds);
+    for (size_t round = 0; round < num_rounds; ++round) {
+      const std::vector<math::RowMajorMatrix<F>>& matrices =
+          matrices_by_round[round];
+      const Points& points = points_by_round[round];
+      OpenedValuesForRound opened_values_for_round(matrices.size());
+      for (size_t matrix_idx = 0; matrix_idx < matrices.size(); ++matrix_idx) {
+        const math::RowMajorMatrix<F>& mat = matrices[matrix_idx];
+        size_t num_rows = static_cast<size_t>(mat.rows());
+        size_t num_cols = static_cast<size_t>(mat.cols());
+        size_t log_num_rows = base::bits::CheckedLog2(num_rows);
+
+        if (reduced_openings[log_num_rows].empty()) {
+          reduced_openings[log_num_rows] =
+              std::vector<ExtF>(num_rows, ExtF::Zero());
+        }
+        std::vector<ExtF>& reduced_opening_for_log_num_rows =
+            reduced_openings[log_num_rows];
+        CHECK_EQ(reduced_opening_for_log_num_rows.size(), num_rows);
+        OpenedValuesForMat opened_values_for_mat(points[matrix_idx].size());
+
+        for (size_t point_idx = 0; point_idx < points[matrix_idx].size();
+             ++point_idx) {
+          const ExtF& point = points[matrix_idx][point_idx];
+          math::RowMajorMatrix<F> block =
```

Comment:
From what I remember when trying to fix this before, we need this to be a copy and not a retrieval of the references. This is because `InterpolateCoset` uses up this `block`, yet the original `mat` should remain unchanged. I believe `Eigen::Block` takes the references of the values instead and doesn't provide a copy.

Reply:
I think `InterpolateCoset()` shouldn't take this as moved, rather as reference!